### PR TITLE
1193 verify endpoint

### PR DESCRIPTION
**Pull request by sunlight7777777**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -159,30 +159,6 @@ data contains the following fields:
 | user         | string | The user's account address                                                                                                    |
 | blockNumber  | number | The block number of the captcha challenge. This is used to verify that the contacted provider was randomly selected on-chain. |
 
-### Add the Procaptcha Widget to your site with React
```
Comment (HughParry) [inline README.md]: Why are you removing this?

File: demos/client-bundle-example/src/jsBundleTest.html [hunk 2]
```diff
@@ -15,7 +15,7 @@
             <div
                 class="procaptcha"
                 data-theme="light"
-                data-sitekey="5HUBceb4Du6dvMA9BiwN5VzUrzUsX9Zp7z7nSR2cC1TCv5jg"
+                data-sitekey="5HGjWAeFDfFCWPsjFQdVV2Msvz2XtMktvgocEZcCj68kUMaw"
```
Comment (HughParry) [inline demos/client-bundle-example/src/jsBundleTest.html]: Please don't change sitekey

File: demos/client-example/tsconfig.json [hunk 3]
```diff
@@ -10,15 +10,9 @@
         {
             "path": "../../packages/procaptcha-react"
         },
-        {
```
Comment (HughParry) [inline demos/client-example/tsconfig.json]: Why is this removed? Procaptcha frictionless is still in the package

File: demos/client-example/tsconfig.json [hunk 4]
```diff
@@ -10,15 +10,9 @@
...
-            "path": "../../packages/procaptcha-frictionless"
-        },
         {
             "path": "../../packages/cli"
         },
-        {
-            "path": "../../packages/types"
```
Comment (HughParry) [inline demos/client-example/tsconfig.json]: Same as above. Why is types removed

File: demos/cypress-shared/cypress/support/commands.ts [hunk 5]
```diff
@@ -43,15 +43,8 @@ function clickIAmHuman(): Cypress.Chainable<Captcha[]> {
             expect(response?.statusCode).to.equal(200)
             expect(response?.body).to.have.property('captchas')
             const captchas = response?.body.captchas.map(({ captcha }: { captcha: CaptchaWithProof }) => captcha)
-            console.log('-----------------------------captchas', captchas, 'length', captchas.length)
             expect(captchas).to.have.lengthOf(2)
             expect(captchas[0]).to.have.property('items')
-            console.log(
```
Comment (HughParry) [inline demos/cypress-shared/cypress/support/commands.ts]: Why are these logs removed from the cypress tests? Presumably they were used for something

File: dev/scripts/src/util/updateEnv.ts [hunk 6]
```diff
@@ -14,32 +14,32 @@
...
 import { at } from '@prosopo/util'
 import { getEnv } from '@prosopo/cli'
+import { glob } from 'glob'
 import dotenv from 'dotenv'
-import fg from 'fast-glob'
 import fs from 'fs'
 import path from 'path'
 
 const ignore = [
-    '**/node_modules/**',
```
Comment (HughParry) [inline dev/scripts/src/util/updateEnv.ts]: Explain this change please

File: dev/scripts/src/util/updateEnv.ts [hunk 7]
```diff
@@ -14,32 +14,32 @@
 import { Logger } from '@prosopo/common'
 import { at } from '@prosopo/util'
 import { getEnv } from '@prosopo/cli'
+import { glob } from 'glob'
 import dotenv from 'dotenv'
-import fg from 'fast-glob'
```
Comment (HughParry) [inline dev/scripts/src/util/updateEnv.ts]: Why changing from fast-glob to glob? I'm not against this I just don't understand why you're changing it

File: packages/contract/src/accounts/getPair.ts [hunk 8]
```diff
@@ -51,7 +51,9 @@ export function getPair(
     if (!secret && account) {
         return keyring.addFromAddress(account)
     } else if (secret) {
+        console.log(secret)
```
Comment (HughParry) [inline packages/contract/src/accounts/getPair.ts]: Don't log secrets

File: packages/contract/src/accounts/getPair.ts [hunk 9]
```diff
@@ -51,7 +51,9 @@ export function getPair(
...
         if (mnemonicValidate(secret)) {
+            console.log(secret)
```
Comment (HughParry) [inline packages/contract/src/accounts/getPair.ts]: Don't log secrets

File: packages/procaptcha-bundle/package.json [hunk 10]
```diff
@@ -42,7 +42,7 @@
         "@prosopo/procaptcha-react": "0.3.39",
         "@prosopo/types": "0.3.39",
         "@prosopo/util": "0.3.39",
-        "react-dom": "^18.3.1"
+        "react-dom": "^18.2.0"
```
Comment (HughParry) [inline packages/procaptcha-bundle/package.json]: Why are you downgrading react-dom?

File: packages/procaptcha-pow/src/Services/SolverService.ts
Comment (HughParry) [inline packages/procaptcha-pow/src/Services/SolverService.ts:1]: Why has this moved from util to procaptcha-pow?

File: packages/provider/src/api/captcha.ts [hunk 11]
```diff
@@ -215,8 +129,13 @@ export function prosopoRouter(env: ProviderEnvironment): Router {
      */
     router.post(ApiPaths.GetPowCaptchaChallenge, async (req, res, next) => {
         try {
-            const { user, dapp } = GetPowCaptchaChallengeRequestBody.parse(req.body)
-
+            const { userAccount, dappAccount } = req.body
+            // Assert that the user and dapp accounts are strings
+            if (typeof userAccount !== 'string' || typeof dappAccount !== 'string') {
```
Comment (HughParry) [inline packages/provider/src/api/captcha.ts]: Please use zod to parse and use ApiParams to build the zod parse.

Check the endpoint for SubmitCaptchaSolution to see how this is done

File: packages/provider/src/tasks/tasks.ts [hunk 12]
```diff
@@ -74,6 +75,8 @@ export class Tasks {
 
     config: ProsopoConfigOutput
 
+    keyring: Keyring
```
Comment (HughParry) [inline packages/provider/src/tasks/tasks.ts]: Why are you adding keyring to Tasks. I can't see where it's used.

File: packages/provider/src/tasks/tasks.ts [hunk 13]
```diff
@@ -341,6 +347,7 @@ export class Tasks {
         const { storedCaptchas, receivedCaptchas, captchaIds } =
             await this.validateReceivedCaptchasAgainstStoredCaptchas(captchas)
         const { tree, commitmentId } = await this.buildTreeAndGetCommitmentId(receivedCaptchas)
+        console.log((await this.contract.methods.getProvider(this.contract.pair.address, {})).value)
```
Comment (HughParry) [inline packages/provider/src/tasks/tasks.ts]: Remove this log please

File: packages/server/src/config.ts [hunk 14]
```diff
@@ -17,20 +17,15 @@ export const getServerConfig = () =>
     ProsopoServerConfigSchema.parse({
         defaultEnvironment: process.env.PROSOPO_DEFAULT_ENVIRONMENT, // enviromental variables
         defaultNetwork: process.env.PROSOPO_DEFAULT_NETWORK,
-        serverUrl: getServerUrl(),
+        serverUrl:
```
Comment (HughParry) [inline packages/server/src/config.ts]: This links to a closed issue - why are we changing this? I don't understand the comment

File: packages/server/src/index.ts [hunk 15]
```diff
@@ -13,11 +13,13 @@
 // limitations under the License.
 import { ProsopoServer } from './server.js'
 export { ProsopoServer } from './server.js'
-export { getServerConfig, getServerUrl } from './config.js'
+export { getServerConfig } from './config.js'
 import { ProsopoServerConfigOutput } from '@prosopo/types'
 import { getPairAsync } from '@prosopo/contract'
 export const PublicProsopoServer = async (config: ProsopoServerConfigOutput) => {
     // if site key is '' then it will burn address
+    console.log('-----------------------')
```
Comment (HughParry) [inline packages/server/src/index.ts]: Remove logs

File: packages/server/src/server.ts [hunk 16]
```diff
@@ -166,22 +175,17 @@ export class ProsopoServer {
...
-            const verifyRecency = await this.verifyRecency(correctCaptchaBlockNumber, maxVerifiedTime)
-            const isHuman = (await contractApi.query.dappOperatorIsHumanUser(user, this.config.solutionThreshold)).value
-                .unwrap()
-                .unwrap()
-            return isHuman && verifyRecency
-        } catch (error) {
-            // if a user is not in the contract it errors, suppress this error and return false
-            return false
-        }
+        const contractApi = await this.getContractApi()
```
Comment (HughParry) [inline packages/server/src/server.ts]: The previous comment says:

// if a user is not in the contract it errors, suppress this error and return false

When did this change?

File: packages/server/src/server.ts [hunk 17]
```diff
@@ -204,14 +208,27 @@ export class ProsopoServer {
         commitmentId?: string,
         maxVerifiedTime = DEFAULT_MAX_VERIFIED_TIME_CACHED
     ) {
-        this.logger.info('Verifying with provider.')
+        const blockNumberString = blockNumber.toString()
+        const dappUserSignature = this.pair?.sign(blockNumberString)
+        if (!dappUserSignature) {
+            throw new Error('Failed to sign the block number')
```
Comment (HughParry) [inline packages/server/src/server.ts]: Please use Prosopo Errors from common. There's a few error types, use the one which is most relevant.

File: packages/server/src/config.ts (hunk 14 above)
Comment (sunlight7777777) [reply packages/server/src/config.ts]: I didn't change this
Should I change this to origin?

File: README.md (hunk 1 above)
Comment (sunlight7777777) [reply README.md]: I added this

File: demos/client-bundle-example/src/jsBundleTest.html (hunk 2 above)
Comment (sunlight7777777) [reply demos/client-bundle-example/src/jsBundleTest.html]: I have done

File: demos/client-example/tsconfig.json (hunk 3 above)
Comment (sunlight7777777) [reply demos/client-example/tsconfig.json]: I have done

File: demos/client-example/tsconfig.json (hunk 4 above)
Comment (sunlight7777777) [reply demos/client-example/tsconfig.json]: I have done

File: demos/cypress-shared/cypress/support/commands.ts (hunk 5 above)
Comment (sunlight7777777) [reply demos/cypress-shared/cypress/support/commands.ts]: I added this

File: packages/provider/src/tasks/tasks.ts (hunk 13 above)
Comment (sunlight7777777) [reply packages/provider/src/tasks/tasks.ts]: I removed

File: packages/server/src/index.ts (hunk 15 above)
Comment (sunlight7777777) [reply packages/server/src/index.ts]: I removed

File: packages/server/src/server.ts (hunk 16 above)
Comment (sunlight7777777) [reply packages/server/src/server.ts]: I am not sure
Should I change this?

File: packages/server/src/server.ts (hunk 17 above)
Comment (sunlight7777777) [reply packages/server/src/server.ts]: throw new ProsopoContractError('CAPTCHA.INVALID_BLOCK_NO', { context: { error: 'Block number not found' } })

File: packages/server/src/server.ts (hunk 16 above)
Comment (sunlight7777777) [reply packages/server/src/server.ts]: I have done

File: packages/server/src/config.ts (hunk 14 above)
Comment (sunlight7777777) [reply packages/server/src/config.ts]: I have done

File: packages/server/src/server.ts (hunk 17 above)
Comment (sunlight7777777) [reply packages/server/src/server.ts]: I am going to add this

File: dev/scripts/src/util/updateEnv.ts (hunk 6 above)
Comment (sunlight7777777) [reply dev/scripts/src/util/updateEnv.ts]: Fixed

File: dev/scripts/src/util/updateEnv.ts (hunk 7 above)
Comment (sunlight7777777) [reply dev/scripts/src/util/updateEnv.ts]: Fixed

File: packages/contract/src/accounts/getPair.ts (hunk 8 above)
Comment (sunlight7777777) [reply packages/contract/src/accounts/getPair.ts]: Fixed

File: packages/contract/src/accounts/getPair.ts (hunk 9 above)
Comment (sunlight7777777) [reply packages/contract/src/accounts/getPair.ts]: Fixed

File: packages/procaptcha-bundle/package.json (hunk 10 above)
Comment (sunlight7777777) [reply packages/procaptcha-bundle/package.json]: Fixed

File: packages/provider/src/tasks/tasks.ts (hunk 12 above)
Comment (sunlight7777777) [reply packages/provider/src/tasks/tasks.ts]: Fixed

File: packages/provider/src/api/captcha.ts (hunk 11 above)
Comment (sunlight7777777) [reply packages/provider/src/api/captcha.ts]: Fixed

File: demos/client-example-server/src/app.ts [hunk 18]
```diff
@@ -31,23 +31,10 @@ export function getEnvFile(filename = '.env', filepath = './') {
     return path.join(filepath, `${filename}.${env}`)
 }
 
-enum ProsopoVerificationType {
```
Comment (forgetso) [inline demos/client-example-server/src/app.ts:34]: Why has this been removed? It's needed to allow the tests to pass in captcha-private

File: demos/client-example-server/src/controllers/auth.ts [hunk 19]
```diff
@@ -35,34 +35,9 @@ function hashPassword(password: string): string {
     return u8aToHex(blake2b(password))
 }
 
-const verify = async (
```
Comment (forgetso) [inline demos/client-example-server/src/controllers/auth.ts:38]: Why has this been removed? I think you need to merge main into your branch.

File: demos/client-example/src/App.tsx [hunk 20]
```diff
@@ -56,11 +55,10 @@ function App(props: AppProps) {
         dappName: 'client-example',
         defaultEnvironment:
             (process.env.PROSOPO_DEFAULT_ENVIRONMENT as EnvironmentTypes) || EnvironmentTypesSchema.enum.development,
-        serverUrl: getServerUrl(),
+        serverUrl: process.env.PROSOPO_SERVER_URL || 'localhost:9228',
```
Comment (forgetso) [inline demos/client-example/src/App.tsx:58]: Please don't change this

File: dev/scripts/src/cli/index.ts [hunk 21]
```diff
@@ -46,23 +46,17 @@ export async function processArgs(args: string[]) {
...
                     .option('deployer', {
                         type: 'string',
-                        demandOption: false,
+                        demand: false,
                         desc: `The account prefix that will deploy the contract. Specifying PROVIDER will cause the 
                         script to look for PROVIDER_JSON in the env file. Specifying DEPLOYER will cause the script to 
                         look for DEPLOYER_JSON in the env file. Defaults to undefined.`,
                         default: undefined,
-                    })
-                    .option('cwd', {
```
Comment (forgetso) [inline dev/scripts/src/cli/index.ts:61]: Please don't change this

File: dev/scripts/src/cli/index.ts [hunk 22]
```diff
@@ -76,8 +70,7 @@ export async function processArgs(args: string[]) {
                     await updateEnvFiles(
                         ['PROSOPO_CONTRACT_ADDRESS', 'NEXT_PUBLIC_PROSOPO_CONTRACT_ADDRESS'],
                         protocolContractAddress.toString(),
-                        log,
```
Comment (forgetso) [inline dev/scripts/src/cli/index.ts:79]: Please don't change this

File: packages/contract/src/accounts/mnemonic.ts [hunk 23]
```diff
@@ -29,12 +29,3 @@ export async function generateMnemonic(keyring?: Keyring, pairType?: KeypairType
     const account = keyring.addFromMnemonic(mnemonic)
     return [mnemonic, account.address]
 }
-
-/** Generate a secret, returning the secret and associated address
```
Comment (forgetso) [inline packages/contract/src/accounts/mnemonic.ts:33]: Please don't change this

File: packages/procaptcha-pow/src/Services/Manager.ts [hunk 24]
```diff
@@ -32,7 +32,7 @@ import { WsProvider } from '@polkadot/rpc-provider/ws'
 import { ContractAbi as abiJson } from '@prosopo/captcha-contract/contract-info'
 import { buildUpdateState, getDefaultEvents } from '@prosopo/procaptcha-common'
 import { sleep } from '@prosopo/procaptcha'
-import { solvePoW } from '@prosopo/util'
+import { solvePoW } from './SolverService.js'
```
Comment (forgetso) [inline packages/procaptcha-pow/src/Services/Manager.ts:35]: Please don't change this

**Review by forgetso — CHANGES_REQUESTED** [review]
@Vengi717 lots of code has been changed for no apparent reason. Please make sure you merge `main` into your branch and incorporate all of the changes into it without simply deleting the code from `main`.

File: packages/server/src/server.ts (hunk 16 above)
Comment (forgetso) [reply packages/server/src/server.ts]: Please don't change this. The error is suppressed for a reason.
